### ISSUE TITLE
Inconsistant and invalid Events Typing using cypress

### ISSUE THREAD

**Issue by avallete** [cypress-io/cypress#5650]
### Current behavior:

Testing a code who used `event instanceof KeyboardEvent`, I've seen some strange behavior, digging up, it seems that there is some strange inconsistencies between the events triggered and also into the way they are sent to the `eventListener`.

### Desired behavior:

Consistent behavior:
- `.type`
- `.trigger({..., type: 'keydown'})`
-  `.get('#element').then(element.dispatchEvent(new KeyboardEvent('keydown', ...)))`

Should all send a KeyboardEvent type event.

And the eventListener:

```js
element.addEventListener('keydown', (event) => (event instanceof KeyboardEvent));
```

Should always return true.

### Steps to reproduce: (app code and test code)

I've tested three different methods: `cy.type`, `cy.trigger`, `cy.get('#element').then( element.dispatchEvent)`.

And two different `eventListener` for each test:

1. One directly into a <script> inside the html page itself
2. One created into the test itself using `cy.get -> addEventListener`

You can find MWE here: https://github.com/avallete/cypress-test-tiny-event-bug

Simply run:

```
npm install
npm run cypress:open

# Then run the test and check the console.log inside the debuging tools.
# Each test is a specific case the 2 first show that event is type Event instead of KeyboardEvent.
# The third test show the inconsistency of the same eventListener if it's declared inside the test or outside (into the html <script>).
```

### Versions

3.6.1, Linux, Chromium 78 / Electron 73

**Comment by avallete** [comment]
The issue also seem to concern for 'mousedown' and MouseEvent, except that `.click` seem to send MouseEvent to one of the two listener. But `.trigger`.

I've added the tests for the MouseEvent to the MWE repo.

**Comment by jennifer-shehane** [comment]
Hey @avallete, I'm trying to narrow down exactly what the issue is from what you provided. Does the below summarize the totality of the issue?

When using `.type()`, `.trigger()` or dispatching an event directly from an element returned from a Cypress command:

- The `mousedown` event that is triggered not an `instanceof MouseEvent`
- The `keydown` event that is triggered in Cypress not an `instanceof KeyboardEvent`

Please clarify if I have missed something else. Thanks.

**Comment by avallete** [comment]
> Hey @avallete, I'm trying to narrow down exactly what the issue is from what you provided. Does the below summarize the totality of the issue?
> 
> When using `.type()`, `.trigger()` or dispatching an event directly from an element returned from a Cypress command:
> 
> * The `mousedown` event that is triggered not an `instanceof MouseEvent`
> * The `keydown` event that is triggered in Cypress not an `instanceof KeyboardEvent`
> 
> Please clarify if I have missed something else. Thanks.

I'll just add from my last comment that:

When using the `.click()` Cypress command, the event that is triggered is not an `instance of MouseEvent`.


But yeah, you have summarized it way better than I did.

**Comment by avallete** [comment]
Investigating a little bit on this issue, I've been able to found what seem to be the root of the behaviour I describe on this issue:

https://github.com/cypress-io/cypress/blob/6749a0edc69a0052497e3a12766723274d78a135/packages/driver/src/cy/keyboard.ts#L892

Replacing the `win['Event']` by `win['KeyboardEvent']` on this line, make the `.type` function to properly send an event which is `instanceof KeyboardEvent` into the tests.

But looking on the comment on top of this line, this fix may lead to some regression on Chrome < 63.

**Comment by sainthkh** [comment]
I tried it with Cypress 4.12.1 + `'Event'` => `'KeyboardEvent'`. But something interesting happened.

`console.log` says it's a `KeyboardEvent`, but the `instanceof KeyboardEvent` is false. 

Here are the screenshots. 

![Screenshot from 2020-08-07 10-33-25](https://user-images.githubusercontent.com/8130013/89601376-52df7f80-d89f-11ea-87ed-a33082304874.png)

Something similar happens to `MouseEvent`.

![Screenshot from 2020-08-07 10-33-16](https://user-images.githubusercontent.com/8130013/89601364-4b1fdb00-d89f-11ea-8de4-a0865ae32d4d.png)

It's a bit hard to figure out what made the difference.

**Comment by avallete** [comment]
I had opened this issue so long ago that I had completely forgot about it.

Related to your comment @sainthkh I've refactored my [MEW](https://github.com/avallete/cypress-test-tiny-event-bug) to:

1. Bump cypress to the latest version
2. Split the different executions contexts of the `instanceof` (inside the running test, inside the html web page script)
3. Rename the tests accordingly to the execution context actually tested.
4. Instead of just logging the result, make the cypress tests actually fail if the `instanceof` is not true in any of the context (since it's the attended behaviour)

If you don't want to clone the MWE repository, here is the videos of the running tests to see what's wrong.

KeyboardEvent tests:

![specjs](https://user-images.githubusercontent.com/8771783/89637726-ff485280-d8ed-11ea-86bc-a88140c88732.gif)

MouseEvent tests:

![specmousejs](https://user-images.githubusercontent.com/8771783/89637748-0b341480-d8ee-11ea-91a6-f426cd7af6eb.gif)


FYI, the expected behaviour for both KeyboardEvent and MouseEvent can be summarize into three assertions who are the following ones:

1. When you add an eventListener to an element you've retrieved inside your test with cypress, and that an event occur, this event should be an `instanceof` (MouseEvent or KeyboardEvent, depending on the test).
2. When an event is dispatched to the html page, the eventListener inside the `<script>` tag of the page, should see an event which is 'instanceof'  (MouseEvent or KeyboardEvent, depending on the test). And so, set the text 'isKeyboardEvent|isMouseEvent' to true inside the '#result' div.
3. Whatever if you use different methods (`.click()`, `.trigger()`, `.type()`, `.get(element).then(elem => elem.dispatchEvent())`) the behaviour should be consistent between all the raised events.

I hope this addition may help to understand more clearly what's going on here.

**Comment by sainthkh** [comment]
Before I go fix `cy.trigger()`, I want to ask this question. When does this consistency necessary? Can I get some real world use cases?

I'm asking this because to solve this problem, we need to add a big table of event name and event object types based on [this](https://developer.mozilla.org/en-US/docs/Web/Events#Standard_events).

**Comment by avallete** [comment]
Hi there,

As you mentioned, the last 3 tests are effectively a combination of the check in both (html script and tests runner) contexts. Which as you say, if all the above tests work, should also work. I do realize know that it may be confusing, sorry for that.

Nice spot about the iframe and the use of win['KeyboardEvent']. It may be worthy to add this trick somewhere into the documentation since this problem should occurs for any `instanceof` or `===` checks on window instantiated objects.

On the question about 'why it is necessary ?':

1. To me it isn't, at first this issue was really about the fact that `.type`  and `.click` wasn't `instanceof` the proper event. I do believe that the `.trigger` issue may even be handled in another issue.
2. Still, I do believe that calling `.trigger('keydown')` or `.type` should produce the same kind of event, and that, according to the MDN, those events should have to proper types (as the ones who would be normally triggered by an user of the tested application.).

**Comment by sainthkh** [comment]
There are 3 problems in this issue:

* `Event` -> `KeyboardEvent: Actually, this is also the solution for the #6125. I need to investigate what's the real problem in that code. And it'll be fixed with it. 
* `cy.window()` issue: I think it should be written in [FAQ](https://docs.cypress.io/faq/questions/using-cypress-faq.html), [cy.window()](https://docs.cypress.io/api/commands/window.html#Syntax), and/or [The Test Runner](https://docs.cypress.io/guides/core-concepts/test-runner.html#Application-Under-Test).
* `trigger` issue: it should be treated separately.

**Comment by jennifer-shehane** [comment]
@sainthkh This seemed to be a use case that someone encountered in their app, where they were checking `instanceof MouseEvent` in their app code, so that cypress didn't behave the same. https://github.com/cypress-io/cypress/issues/8137 

But yah, as you mentioned, this trigger thing may be a separate issue. Feel free to reopen that issue if it's the case.

**Comment by sainthkh** [comment]
@jennifer-shehane I guess so. After #8255 is done, I'll research it more deeply.

**Comment by tobmaster** [comment]
I am not very experienced in the cypress codebase yet. But shouldn't the action "type" emit a [KeyboardEvent](https://www.w3.org/TR/2003/NOTE-DOM-Level-3-Events-20031107/events.html#Events-KeyboardEvents-Interfaces) instead of an HTMLEvent?
https://github.com/cypress-io/cypress/blob/093f3e7505096e24b2d49d53dbf2e9b3a31dd6fe/packages/driver/src/cy/commands/actions/type.js#L222

That way [getModifierState()](https://www.w3.org/TR/2003/NOTE-DOM-Level-3-Events-20031107/events.html#Events-KeyboardEvent-getModifierState) could work correctly?
The modifieres still need to be set on the event I assume.

We currently have this problem in the project where we handle modifiers and it fails cause the method is not there. And including a workaround for cypress in the application code is kinda awkward. 
Maybe I am wrong here but this seems to be connected to this issue.

**Comment by tobmaster** [comment]
Ah I now also read all of #8255 So I assume my comment is obsolete then. Sorry for the noise

**Comment by anp** [comment]
> Before I go fix `cy.trigger()`, I want to ask this question. When does this consistency necessary? Can I get some real world use cases?

I'm running into this issue because I have some Rust code which uses [typed bindings](https://docs.rs/web-sys/0.3.44/web_sys/index.html) to DOM APIs. The bindings provide a number of type definitions generated from WebIDL, and it's possible to [cast between JS types](https://docs.rs/wasm-bindgen/0.2.67/wasm_bindgen/trait.JsCast.html). By default those casts use `instanceof` checks before returning to catch common bugs.

I have [event handler helpers](https://github.com/anp/moxie/blob/main/dom/augdom/src/webdom.rs#L21-L31) that take a typed callback and perform the conversion under the hood for the user. Elsewhere there are definitions mapping strings like `"keydown"` to types so that the callbacks always have the correct type. This works quite nicely "in production" so far but breaks when using cypress to enter text due to this issue.

I've had mixed results with removing the checks as it seems the bindings may rely on prototypes for doing their property lookups. The easiest fix that'd allow me to use cypress would be to emit an event with the correct prototype in its chain.
